core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -84,19 +76,65 @@ public void cleanFiles(TableMetadata beforeExpiration, TableMetadata afterExpira
     deleteFiles(manifestListsToDelete, "manifest list");
   }
 
-  private Set<ManifestFile> readManifests(Set<Snapshot> snapshots) {
-    Set<ManifestFile> manifestFiles = Sets.newHashSet();
-    for (Snapshot snapshot : snapshots) {
-      try (CloseableIterable<ManifestFile> manifestFilesForSnapshot = readManifestFiles(snapshot)) {
-        for (ManifestFile manifestFile : manifestFilesForSnapshot) {
-          manifestFiles.add(manifestFile.copy());
-        }
-      } catch (IOException e) {
-        throw new RuntimeIOException(
-            e, "Failed to close manifest list: %s", snapshot.manifestListLocation());
-      }
+  private Set<ManifestFile> currentManifests(
+      Set<Snapshot> snapshots, Set<ManifestFile> manifestsToDelete) {
+    Set<ManifestFile> currentManifests = ConcurrentHashMap.newKeySet();
+    if (manifestsToDelete.isEmpty()) {
+      return currentManifests;
     }
 
+    Tasks.foreach(snapshots)
+        .retry(3)
+        .stopOnFailure()
+        .throwFailureWhenFinished()
+        .executeWith(planExecutorService)
+        .onFailure(
+            (snapshot, exc) ->
+                LOG.warn(
+                    "Failed to determine manifests for snapshot {}", snapshot.snapshotId(), exc))
+        .run(
+            snapshot -> {
+              try (CloseableIterable<ManifestFile> manifestFiles = readManifests(snapshot)) {
+                for (ManifestFile manifestFile : manifestFiles) {
+                  manifestsToDelete.remove(manifestFile);
+                  if (manifestsToDelete.isEmpty()) {
+                    return;
+                  }
+
+                  currentManifests.add(manifestFile.copy());
```

Comment:
This seems potentially incorrect to me. It seems like we could get different results in `currentManifests` based on whether or not there are more manifests left to delete.

Specifically, if some manifest file M is the last remaining manifest to delete, then it won't be included in `currentManifests`.

However, if the snapshots are evaluated in an ordering such that M is considered while there are still manigests left to delete, then M _will be_ included in `currentManifests`.

This behavior that depends on ordering is suspicious to me, though I was admittedly not involved in the previous PR review.

Reply:
I think [this](https://github.com/apache/iceberg/pull/5981/files#r996389820) should explain the rationale for why there is correctness, but do please let me know if you still have concerns. This procedure should certainly guarantee we're deleting the expected files